src/Carbon/CarbonPeriod.php

```diff
@@ -0,0 +1,1623 @@
+     *
+     * @see https://bugs.php.net/bug.php?id=72255
+     * @see https://bugs.php.net/bug.php?id=74274
+     * @see https://wiki.php.net/rfc/datetime_and_daylight_saving_time
+     *
+     * @return void
+     */
+    protected function initializeCurrentDate()
+    {
+        $this->current = $this->startDate->copy();
+
+        if (static::intervalHasTime($this->dateInterval)) {
+            $this->timezone = $this->current->getTimezone();
+
+            $this->current->setTimezone('UTC');
+        } else {
+            $this->timezone = null;
+        }
+    }
+
+    /**
+     * Keep incrementing the current date until a valid date is found or the iteration is ended.
+     *
+     * @throws \RuntimeException
+     *
+     * @return void
+     */
+    protected function incrementCurrentDateUntilValid()
+    {
+        $attempts = 0;
+
+        do {
+            $this->current->add($this->dateInterval);
+
+            $this->validationResult = null;
+
+            if (++$attempts > static::NEXT_MAX_ATTEMPTS) {
+                throw new RuntimeException('Could not find next valid date.');
+            }
+        } while ($this->validateCurrentDate() === false);
+    }
+
+    /**
+     * Set current date from the iteration results and return true on success.
+     *
+     * @param int $key
+     *
+     * @return bool
+     */
+    protected function setCurrentDateFromCache($key)
+    {
+        if ($this->parametersChanged || $this->isCacheDisabled()) {
+            return false;
+        }
+
+        if ($key < count($this->iterationResults)) {
+            $this->validationResult = true;
+
+            $this->key = $key;
+            $this->current = $this->iterationResults[$key]->copy();
+
+            return true;
+        }
+
+        if ($this->iterationCompleted) {
+            $this->validationResult = static::END_ITERATION;
+
+            return true;
+        }
+
+        return false;
+    }
+
+    /**
+     * Format the date period as ISO 8601.
+     *
+     * @return string
+     */
+    public function toIso8601String()
+    {
+        $parts = array();
+
+        if ($this->recurrences !== null) {
+            $parts[] = 'R'.$this->recurrences;
+        }
+
+        if ($this->startDate !== null) {
+            $parts[] = $this->startDate->toIso8601String();
+        }
+
+        $parts[] = $this->dateInterval->spec();
+
+        if ($this->endDate !== null) {
+            $parts[] = $this->endDate->toIso8601String();
+        }
+
+        return implode('/', $parts);
+    }
+
+    /**
+     * Convert the date period into a string.
+     *
+     * @return string
+     */
+    public function toString()
+    {
+        $translator = Carbon::getTranslator();
+
+        $parts = array();
+
+        if ($this->startDate && !$this->startDate->isStartOfDay() || $this->endDate && !$this->endDate->isStartOfDay()) {
+            $format = 'Y-m-d H:i:s';
+        } else {
+            $format = 'Y-m-d';
+        }
```

Comment:
Ternary for just the hour part would be cleaner.

Reply:
Not a big fan of ternaries with long conditions here: `$format = 'Y-m-d'.($this->startDate && !$this->startDate->isStartOfDay() || $this->endDate && !$this->endDate->isStartOfDay() ? ' H:i:s' : '')` doesn't look very clean. I was thinking to make into a variable first or a separate method, but... why avoid `if .. else` so much?

Reply:
I just never use `else` in my own project. Check https://github.com/phug-php, it's a very complex project, a lot more code than in Carbon, and it contains no single `else` statement. I did not invented this rule it's a PHPMD wide-used rule: https://phpmd.org/rules/cleancode.html#elseexpression

The meaning of this is: a `else` statement create a block when you can add treatment, it opens the door too complexity. If it's really needed, it deserves a dedicated method. If it cearly match the ternary pattern (like here), I mean: an assignation to the same variable in `if` and `else`, then ternary has the advantage to just be an operator like `+`, `-`, etc. it's a simple operation, not block, statement and stuff that more likely will multiply paths.

But if the ternary is long, there is no particular reason to write it on one line:
```php
$format = 'Y-m-d'.(
  $this->startDate && !$this->startDate->isStartOfDay() || $this->endDate && !$this->endDate->isStartOfDay()
    ? ' H:i:s'
    : ''
);
```